### ISSUE TITLE
How to calculate the loss

### ISSUE THREAD

**Issue by orydatadudes** [lllyasviel/ControlNet#60]
i didn't fully understand how the loss is calculated 

regular diffusion models takes input image adding noise and during backward stage we learn how to undo that noise 
the loss calculate base on how good we learn that noise for every stage in T 

so what i don't understand is what the part of the target image ? ( we add the noise to input image )

that is what says in the paper 

![image](https://user-images.githubusercontent.com/64726228/219061337-317f971b-ce74-4b6d-b3fb-5d65897f580f.png)

I also not sure what is it "task-specific conditions cf" in the loss , is it the target image ?

thanks

**Comment by xiankgx** [comment]
This paper is about adding additional condition to an existing text-conditioned generative model. In the paper, there are many tasks mentioned (e.g. controlling generation with canny-edge, controlling generation with hough-line, controlling generation with user scribble, etc). For canny-edge input, c_f is the canny-edge image obtained from the ground truth image.

**Comment by orydatadudes** [comment]
so it c_f is the ground truth (canny-edge for that task) and your model learn how to make prediction with that input how you can use that model for another/test image ?

**Comment by xiankgx** [comment]
> so it c_f is the ground truth (canny-edge for that task) and your model learn how to make prediction with that input how you can use that model for another/test image ?

During training, there is a set of such input and targets, and we minimize the loss on this dataset. The ultimate goal of all neural net training is for it to generalize (meaning also work on another test image that it may not have seen during training).

**Comment by orydatadudes** [comment]
thanks for the answer but it's still not fully clear for me 

for canny edge missing, during training the input to the model is [(canny_edge_1, target_image_1,**text prompt)**...()]
canny_edge_1 - is an image with only black & white edges 
target_image - the image that the canny_edge image extracted from 
text - prompt that describe the canny_edge_1 image (which is also the same description for the target_image)

now target_image + text area are used as "hint" to the backward process 
that mean the diffusion process add noise to the **canny_edge_1** ?
that also mean  that for new image i can generate random noise, then add these "hint" (target_image + text) and get new canny edge to my new image & prompt

if that is true , that also mean that the function the learn how to reverse the noise should **NOT** get the canny_edge_1 image but the target_image which is different  from the formula in the paper 

i really get confuse and need some help to figure it out

**Comment by xiankgx** [comment]
The diffusion process NOT add noise to the hint (canny edge) input, it adds noise to the target image. Just like how the diffusion process also do NOT add noise to the text. When comparing to standard SD, there is no change in how the forward and reverse diffusion process works.

What is changed is that new inputs are created (hint input) to influence the denoising process via the SD Unet.

The timesteps, text conditioning, and canny edge input is input to the Controlnet which is used to control/modify the SD Unet behavior.

**Comment by xiankgx** [comment]
To make this easier, just imagine we replace the Unet in SD with a Unet that can take in additional input (hint input).

**Comment by xiankgx** [comment]
![image](https://user-images.githubusercontent.com/4113258/220589928-9e7ad08c-6794-4ce1-90b8-6e9b0dfbc76c.png)

Consider this image, to go from x_T to x_T-1, in normal SD, a Unet is used, and this time variable t, text conditioning data from CLIP is inputted to this Unet.

Now, with ControlNet, this Unet is modified, and this time in addition to time variable t, text conditiioning data from CLIP, it also takes in edge map data.

**Comment by orydatadudes** [comment]
thank you for your detailed response 
 for making canny edge mission when you say  " it adds noise to the **target image**." 
target image  is that kind of images?

![image](https://user-images.githubusercontent.com/64726228/220960368-63e3cde2-484d-4411-9002-1d8268233217.png)


or that ?

![image](https://user-images.githubusercontent.com/64726228/220960487-283709b9-e647-407e-a78e-65243e286f05.png)

**Comment by xiankgx** [comment]
> 
> ![image](https://user-images.githubusercontent.com/64726228/220960368-63e3cde2-484d-4411-9002-1d8268233217.png)
>

The target or label image is what you want the model to predict.
